File: comm.go
// go build comm.go
// ./comm
package main

import (
	//"crypto/md5"
	//"errors"
	"flag"
	"fmt"
	"io"
	"log"
	"math" // 使用标准库的math包进行三角函数计算
	"os"
	"os/signal"
	"reflect"
	"syscall"

	//"mime/multipart"
	//"context"
	"encoding/json" // 用于解析和返回 JSON
	"net/http"
	"strconv"
	"strings"
	"sync"

	_ "net/http/pprof"

	"github.com/quic-go/quic-go"
	"github.com/quic-go/quic-go/http3"

	// "github.com/quic-go/quic-go/internal/testdata"
	"github.com/quic-go/quic-go/qlog"
	// 屏蔽Redis依赖
	//"github.com/redis/go-redis/v9"
	//"github.com/quic-go/quic-go/logging"
)

type binds []string

// 定义数据传输格式，服务端客户端须保持一致
type Data struct {
	Name       string    `json:"Name"`
	IP         string    `json:"IP"`
	Port       int       `json:"Port"`
	X          float64   `json:"X"`
	Y          float64   `json:"Y"`
	Psi        float64   `json:"Psi"`
	Stop_label bool      `json:"Stop_label"`
	Req_Resp   bool      `json:"Req_Resp"`
	V          float64   `json:"V"`
	W          float64   `json:"W"`
	Path_Param []float64 `json:"Path_Param"` // 等价于C++中的std::vector<double>
}

func (b binds) String() string {
	return strings.Join(b, ",")
}

func (b *binds) Set(v string) error {
	*b = strings.Split(v, ",")
	return nil
}

// Size is needed by the /demo/upload handler to determine the size of the uploaded file
// type Size interface {
//  Size() int64
// }

// 在结构体中存储数据，用于模拟Redis
var mockDataStore sync.Map

// See https://en.wikipedia.org/wiki/Lehmer_random_number_generator
func generatePRData(l int) []byte {
	res := make([]byte, l)
	seed := uint64(1)
	for i := 0; i < l; i++ {
		seed = seed * 48271 % 2147483647
		res[i] = byte(seed)
	}
	return res
}

// 模拟删除键的函数
func deleteKeys() {
	log.Println("Deleting mock data store keys...")
	mockDataStore.Range(func(key, value interface{}) bool {
		mockDataStore.Delete(key)
		return true
	})
	log.Println("Mock data store cleared")
}

// generateStraightPath 根据当前位置、朝向生成向斜前方的直线轨迹
// 参数：
//   currentX, currentY - 当前位置坐标
//   heading - 车辆朝向（弧度制，0度为x轴正方向）
//   pointCount - 需要生成的轨迹点数量
// 返回值：
//   格式为 [X1, Y1, X2, Y2, ...] 的轨迹点切片，每个点占两个元素
func generateStraightPath(currentX, currentY, heading float64, pointCount int) []float64 {
	// 创建轨迹点数组，每个点需要2个float64值（X和Y）
	path := make([]float64, pointCount*2)
	
	// 设置起始点为当前位置
	path[0] = currentX
	path[1] = currentY
	
	// 计算每个点之间的步长，使用固定步长使轨迹呈直线
	stepSize := 1.0 // 可以根据需要调整步长
	
	// 从第二个点开始计算
	for i := 1; i < pointCount; i++ {
		// 计算沿朝向方向的偏移量
		stepDistance := float64(i) * stepSize
		
		// 根据朝向计算新点的坐标
		// 朝向为0度时，车辆朝向x轴正方向
		// 角度增加时，逆时针旋转
		xOffset := stepDistance * math.Cos(heading)
		yOffset := stepDistance * math.Sin(heading)
		
		// 设置轨迹点
		path[2*i] = currentX + xOffset
		path[2*i+1] = currentY + yOffset
	}
	
	return path
}

func setupHandler(www string) http.Handler {

	mux := http.NewServeMux()
	// ctx 不再需要，因为我们已经屏蔽了Redis依赖

	if len(www) > 0 {
		mux.Handle("/", http.FileServer(http.Dir(www)))
	} else {
		mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
			fmt.Printf("%#v\n", r)
			const maxSize = 1 << 30 // 1 GB
			num, err := strconv.ParseInt(strings.ReplaceAll(r.RequestURI, "/", ""), 10, 64)
			if err != nil || num <= 0 || num > maxSize {
				w.WriteHeader(400)
				return
			}
			w.Write(generatePRData(int(num)))
		})
	}

	mux.HandleFunc("/demo/hash", func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.Error(w, "Only POST method is supported", http.StatusMethodNotAllowed)
			return
		}
		//解析请求体中的 JSON 数据为struct
		var message Data

		err := json.NewDecoder(r.Body).Decode(&message)
		if err != nil {
			http.Error(w, "Invalid JSON data", http.StatusBadRequest)
			fmt.Printf("Error decoding JSON: %v\n", err)
			return
		}
		fmt.Printf("message.x: %v\n", message.X)

		// 模拟存储数据，将数据存入map
		value := reflect.ValueOf(message)
		typeOf := reflect.TypeOf(message)
		log.Println("Storing data in mock store:")
		for i := 0; i < value.NumField(); i++ {
			fieldName := typeOf.Field(i).Name
			fieldValue := value.Field(i).Interface()
			log.Printf("Field '%s': %v\n", fieldName, fieldValue)
		}

		// 模拟返回数据
		var response Data = message
		// 设置一些模拟的返回值
		response.V = 1.0
		response.W = 0.5

		// 返回相同的 JSON 数据
		w.Header().Set("Content-Type", "application/json")
		err = json.NewEncoder(w).Encode(response)
		if err != nil {
			http.Error(w, "Failed to encode response", http.StatusInternalServerError)
			fmt.Printf("Error encoding JSON: %v\n", err)
		}

	})

	mux.HandleFunc("/demo/string", func(w http.ResponseWriter, r *http.Request) {

		// 只接受 POST 请求
		if r.Method != http.MethodPost {
			http.Error(w, "Only POST method is supported", http.StatusMethodNotAllowed)
			return
		}

		// 读取请求体并解析为字节切片
		bodyBytes, err := io.ReadAll(r.Body)
		if err != nil {
			http.Error(w, "Failed to read request body", http.StatusBadRequest)
			return
		}

		// 解析 JSON 数据为结构体
		var bodyData Data
		var check_command_result Data
		err = json.Unmarshal(bodyBytes, &bodyData)
		if err != nil {
			http.Error(w, "Invalid JSON data", http.StatusBadRequest)
			fmt.Printf("Error decoding JSON: %v\n", err)
			return
		}

		fmt.Printf("Req from: %v\n", bodyData.Name)

		// 模拟在clouder_list中增加对象
		log.Printf("Processing request from: %v\n", bodyData.Name)
		// 检查是否已存在
		listKey := "clouder_list"
		if listData, ok := mockDataStore.Load(listKey); ok {
			if list, ok := listData.(map[string]bool); ok {
				if !list[bodyData.Name] {
					list[bodyData.Name] = true
					mockDataStore.Store(listKey, list)
					fmt.Printf("Add '%s' to clouder_list\n", bodyData.Name)
				}
			}
		} else {
			// 创建新的列表
			newList := map[string]bool{bodyData.Name: true}
			mockDataStore.Store(listKey, newList)
			fmt.Printf("Add '%s' to clouder_list\n", bodyData.Name)
		}

		// 模拟将请求body字符串存入存储，键名为 bodyData.Name
		bodyString := string(bodyBytes)
		mockDataStore.Store(bodyData.Name, bodyString)
		fmt.Println("Request Body:", bodyString)

		// 模拟获取键"carX-c"command的值
		commandKey := bodyData.Name + "-c"
		carcommandData, exists := mockDataStore.Load(commandKey)
		if !exists {
			// 如果键不存在，设置键值
			mockDataStore.Store(commandKey, bodyString)
			fmt.Printf("%s First connection\n", bodyData.Name)
		} else {
			// 如果键存在，打印命令
			carcommand := carcommandData.(string)
			json.Unmarshal([]byte(carcommand), &check_command_result)
			//检查Req_Resp是否为true，将其重置为false
			if check_command_result.Req_Resp {
				log.Printf("Command: %s\n", carcommand)
				check_command_result.Req_Resp = false
				command_ready_label_false, _ := json.Marshal(check_command_result)
				// 将更新后的值存回模拟存储
				mockDataStore.Store(commandKey, string(command_ready_label_false))
			} else {
				// 模拟检查更新，但不实际等待，直接返回当前数据
				log.Println("No new command available, using current data")
			}
			// 处理完成
		}
		//获取键"car"的值
		// car, err := rdb.Get(ctx, "car").Result()
		// if err != nil {
		// 	panic(err)
		// }
		// fmt.Println("car:",car)
		//Incr"car"的值
		// newValue, err := rdb.Incr(ctx, "car").Result()
		// if err != nil {
		//     fmt.Println("Failed to increment 'car':", err)
		//     return
		// }
		//在carlist集合中增加对象car1
		// err = rdb.SAdd(ctx, "carlist", "car"+car).Err()
		// if err != nil {
		// 	fmt.Println("Failed to SADD 'carlist':", err)
		// 	return
		// }
		//获取carlist集合中所有对象
		// carlist, err := rdb.SMembers(ctx, "carlist").Result()
		// if err != nil {
		// 	fmt.Println("Failed to get members of 'carlist':", err)
		// 	return
		// }
		// fmt.Println("car:",carlist)

		//json数据的反序列化
		// var response Data
		// err = json.Unmarshal([]byte(carcommand), &response)
		//如果JSON结构不固定，可以使用 map[string]interface{} 作为动态结构：
		// var jsonData map[string]interface{}
		// err = json.Unmarshal([]byte(bodyString), &jsonData)
		// if err != nil {
		// 	fmt.Println("Error decoding JSON:", err)
		// 	return
		// }
		//fmt.Printf("Decoded JSON as map: %v\n", jsonData)

		// 处理Path_Param，生成向斜前方的直线轨迹
		// 首先解析响应数据到结构体
		var response Data
		if carcommandData, exists := mockDataStore.Load(commandKey); exists {
			json.Unmarshal([]byte(carcommandData.(string)), &response)
		} else {
			response = bodyData // 如果没有存储的命令，使用请求数据
		}
		
		// 生成轨迹点：从当前位置向斜前方延伸的直线
		response.Path_Param = generateStraightPath(bodyData.X, bodyData.Y, bodyData.Psi, 20)
		
		// 记录生成的轨迹信息
		fmt.Printf("Generated path with %d points, starting at (%.2f, %.2f), direction: %.2f radians\n", 
			len(response.Path_Param)/2, bodyData.X, bodyData.Y, bodyData.Psi)
		
		// 将更新后的响应结构体序列化为JSON并返回
		w.Header().Set("Content-Type", "application/json")
		err = json.NewEncoder(w).Encode(response)
		if err != nil {
			http.Error(w, "Failed to encode response", http.StatusInternalServerError)
			fmt.Printf("Error encoding JSON: %v\n", err)
			return
		}

		// 将结构体数据response变量 写入http响应，返回JSON 数据
		// w.Header().Set("Content-Type", "application/json")
		// err = json.NewEncoder(w).Encode(response)
		// if err != nil {
		// 	http.Error(w, "Failed to encode response", http.StatusInternalServerError)
		// 	fmt.Printf("Error encoding JSON: %v\n", err)
		// }

	})
	return mux
}

func main() {

	// 屏蔽Redis连接
	// rdb := redis.NewClient(&redis.Options{
	// 	Addr:     "192.168.100.51:30079",
	// 	Password: "", // 没有密码，默认值
	// 	DB:       0,  // 默认DB 0
	// })

	// defer profile.Start().Stop()
	go func() {
		log.Println(http.ListenAndServe("localhost:6060", nil))
	}()
	// runtime.SetBlockProfileRate(1)

	bs := binds{}
	flag.Var(&bs, "bind", "bind to")
	www := flag.String("www", "", "www data")
	tcp := flag.Bool("tcp", false, "also listen on TCP")
	key := flag.String("key", "", "TLS key (requires -cert option)")
	cert := flag.String("cert", "", "TLS certificate (requires -key option)")
	flag.Parse()

	if len(bs) == 0 {
		bs = binds{"0.0.0.0:6121"}
	}

	handler := setupHandler(*www)

	var wg sync.WaitGroup
	wg.Add(len(bs))

	var certFile, keyFile string
	if *key != "" && *cert != "" {
		keyFile = *key
		certFile = *cert
	} else {
		certFile = "certpath/cert.pem"
		keyFile = "certpath/priv.key"
	}

	for _, b := range bs {
		fmt.Println("listening on", b)
		bCap := b
		go func() {
			var err error
			if *tcp {
				err = http3.ListenAndServeTLS(bCap, certFile, keyFile, handler)
			} else {
				server := http3.Server{
					Handler: handler,
					Addr:    bCap,
					QUICConfig: &quic.Config{
						Tracer: qlog.DefaultConnectionTracer,
					},
				}
				err = server.ListenAndServeTLS(certFile, keyFile)
			}
			if err != nil {
				fmt.Println(err)
			}
			wg.Done()
		}()

		// 使用 os/signal 来捕获 SIGINT 和 SIGTERM 信号（如 Ctrl+C 或 kill）
		signalChan := make(chan os.Signal, 1)
		signal.Notify(signalChan, syscall.SIGINT, syscall.SIGTERM)

		// 等待退出信号
		sig := <-signalChan
		log.Printf("Received signal: %s. Shutting down...\n", sig)

		// 在退出前删除模拟存储中的相关键
		//deleteKeys()

		// 执行优雅退出
		log.Println("Exiting...")
		os.Exit(0) // 退出程序

	}
	wg.Wait()
}

File: comm.py
# community.py
import asyncio
import json
import logging
import math
import os
import signal
import sys
from collections import defaultdict
from typing import Dict, List, Optional

from aioquic.asyncio import QuicConnectionProtocol, serve
from aioquic.h3.connection import H3_ALPN, H3Connection
from aioquic.h3.events import (
    DataReceived,
    H3Event,
    HeadersReceived,
)
from aioquic.quic.configuration import QuicConfiguration
from aioquic.quic.events import QuicEvent
from aioquic.tls import SessionTicket

# 设置日志
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(name)s %(message)s",
)
logger = logging.getLogger("community")

# 模拟Go代码中的数据结构
class Data:
    def __init__(self, **kwargs):
        self.Name = kwargs.get("Name", "")
        self.IP = kwargs.get("IP", "")
        self.Port = kwargs.get("Port", 0)
        self.X = kwargs.get("X", 0.0)
        self.Y = kwargs.get("Y", 0.0)
        self.Psi = kwargs.get("Psi", 0.0)
        self.Stop_label = kwargs.get("Stop_label", False)
        self.Req_Resp = kwargs.get("Req_Resp", False)
        self.V = kwargs.get("V", 0.0)
        self.W = kwargs.get("W", 0.0)
        self.Path_Param = kwargs.get("Path_Param", [])

    def to_dict(self):
        return {
            "Name": self.Name,
            "IP": self.IP,
            "Port": self.Port,
            "X": self.X,
            "Y": self.Y,
            "Psi": self.Psi,
            "Stop_label": self.Stop_label,
            "Req_Resp": self.Req_Resp,
            "V": self.V,
            "W": self.W,
            "Path_Param": self.Path_Param
        }

# 模拟Go代码中的mockDataStore
mock_data_store = defaultdict(dict)

class HttpRequestHandler:
    def __init__(self):
        self.method = ""
        self.path = ""
        self.headers = {}
        self.body = b""
        self.stream_ended = False

    def add_headers(self, headers):
        """添加请求头"""
        for header, value in headers:
            if header == b":method":
                self.method = value.decode()
            elif header == b":path":
                self.path = value.decode()
            else:
                # 存储其他头信息
                self.headers[header.decode()] = value.decode()

    def add_data(self, data: bytes, stream_ended: bool):
        """添加请求体数据"""
        self.body += data
        self.stream_ended = stream_ended

    def is_complete(self):
        """检查请求是否完整"""
        return self.stream_ended

    async def handle_request(self, protocol, stream_id):
        """处理HTTP请求"""
        if not self.is_complete():
            logger.error(f"Request not complete for stream {stream_id}")
            return

        if self.method != "POST":
            await self.send_error(protocol, stream_id, 405, "Only POST method is supported")
            return

        if self.path == "/demo/hash":
            await self.handle_demo_hash(protocol, stream_id)
        elif self.path == "/demo/string":
            await self.handle_demo_string(protocol, stream_id)
        else:
            await self.send_error(protocol, stream_id, 404, "Not Found")

    async def handle_demo_hash(self, protocol, stream_id):
        """处理 /demo/hash 端点"""
        try:
            # 解析JSON数据
            message_data = json.loads(self.body.decode())
            message = Data(**message_data)
            
            logger.info(f"message.x: {message.X}")
            logger.info("Storing data in mock store:")
            
            # 记录所有字段
            for field_name, field_value in message.to_dict().items():
                logger.info(f"Field '{field_name}': {field_value}")

            # 创建响应数据
            response = Data(**message_data)
            response.V = 1.0
            response.W = 0.5

            # 发送JSON响应
            await self.send_json_response(protocol, stream_id, response.to_dict())
            
        except json.JSONDecodeError as e:
            await self.send_error(protocol, stream_id, 400, "Invalid JSON data")
        except Exception as e:
            logger.error(f"Error in /demo/hash: {e}")
            await self.send_error(protocol, stream_id, 500, "Internal Server Error")

    async def handle_demo_string(self, protocol, stream_id):
        """处理 /demo/string 端点"""
        try:
            # 解析JSON数据
            body_data = json.loads(self.body.decode())
            body_obj = Data(**body_data)
            
            logger.info(f"Req from: {body_obj.Name}")

            # 模拟在clouder_list中增加对象
            list_key = "clouder_list"
            if list_key in mock_data_store:
                clouder_list = mock_data_store[list_key]
                if body_obj.Name not in clouder_list:
                    clouder_list[body_obj.Name] = True
                    logger.info(f"Add '{body_obj.Name}' to clouder_list")
            else:
                mock_data_store[list_key] = {body_obj.Name: True}
                logger.info(f"Add '{body_obj.Name}' to clouder_list")

            # 存储请求体
            mock_data_store[body_obj.Name] = self.body.decode()
            logger.info(f"Request Body: {self.body.decode()}")

            # 检查命令键
            command_key = f"{body_obj.Name}-c"
            check_command_result = Data()
            
            if command_key in mock_data_store:
                car_command = mock_data_store[command_key]
                check_command_result = Data(**json.loads(car_command))
                
                if check_command_result.Req_Resp:
                    logger.info(f"Command: {car_command}")
                    check_command_result.Req_Resp = False
                    command_ready_label_false = json.dumps(check_command_result.to_dict())
                    mock_data_store[command_key] = command_ready_label_false
                else:
                    logger.info("No new command available, using current data")
            else:
                mock_data_store[command_key] = self.body.decode()
                logger.info(f"{body_obj.Name} First connection")

            # 准备响应数据
            if command_key in mock_data_store:
                response_data = json.loads(mock_data_store[command_key])
                response = Data(**response_data)
            else:
                response = body_obj

            # 生成路径参数
            response.Path_Param = self.generate_straight_path(
                body_obj.X, body_obj.Y, body_obj.Psi, 20
            )
            
            logger.info(f"Generated path with {len(response.Path_Param)//2} points, "
                       f"starting at ({body_obj.X:.2f}, {body_obj.Y:.2f}), "
                       f"direction: {body_obj.Psi:.2f} radians")

            # 发送响应
            await self.send_json_response(protocol, stream_id, response.to_dict())

        except json.JSONDecodeError as e:
            await self.send_error(protocol, stream_id, 400, "Invalid JSON data")
        except Exception as e:
            logger.error(f"Error in /demo/string: {e}")
            await self.send_error(protocol, stream_id, 500, "Internal Server Error")

    def generate_straight_path(self, current_x: float, current_y: float, 
                             heading: float, point_count: int) -> List[float]:
        """生成直线路径，与Go版本保持一致"""
        path = [0.0] * (point_count * 2)
        path[0] = current_x
        path[1] = current_y
        
        step_size = 1.0
        
        for i in range(1, point_count):
            step_distance = float(i) * step_size
            x_offset = step_distance * math.cos(heading)
            y_offset = step_distance * math.sin(heading)
            
            path[2*i] = current_x + x_offset
            path[2*i+1] = current_y + y_offset
        
        return path

    async def send_json_response(self, protocol, stream_id, data: dict):
        """发送JSON响应"""
        response_body = json.dumps(data).encode()
        headers = [
            (b":status", b"200"),
            (b"content-type", b"application/json"),
            (b"content-length", str(len(response_body)).encode()),
        ]
        
        protocol._http.send_headers(
            stream_id=stream_id,
            headers=headers,
        )
        protocol._http.send_data(
            stream_id=stream_id,
            data=response_body,
            end_stream=True
        )

    async def send_error(self, protocol, stream_id, status_code: int, message: str):
        """发送错误响应"""
        error_data = {"error": message}
        response_body = json.dumps(error_data).encode()
        headers = [
            (b":status", str(status_code).encode()),
            (b"content-type", b"application/json"),
            (b"content-length", str(len(response_body)).encode()),
        ]
        
        protocol._http.send_headers(
            stream_id=stream_id,
            headers=headers,
        )
        protocol._http.send_data(
            stream_id=stream_id,
            data=response_body,
            end_stream=True
        )

class Http3ServerProtocol(QuicConnectionProtocol):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._http = None
        self._handlers = {}  # 存储每个流的处理器

    def quic_event_received(self, event: QuicEvent):
        if self._http is None:
            self._http = H3Connection(self._quic)
        
        # 将QUIC事件传递给HTTP/3层
        for h3_event in self._http.handle_event(event):
            self.h3_event_received(h3_event)

    def h3_event_received(self, event: H3Event):
        if isinstance(event, HeadersReceived):
            self.handle_headers(event)
        elif isinstance(event, DataReceived):
            self.handle_data(event)

    def handle_headers(self, event: HeadersReceived):
        """处理HTTP头"""
        stream_id = event.stream_id
        
        # 为新的流创建处理器
        if stream_id not in self._handlers:
            self._handlers[stream_id] = HttpRequestHandler()
        
        handler = self._handlers[stream_id]
        handler.add_headers(event.headers)
        
        # 如果流已经结束（没有请求体），立即处理请求
        if event.stream_ended:
            handler.add_data(b"", True)
            asyncio.create_task(self.process_request(stream_id))

    def handle_data(self, event: DataReceived):
        """处理HTTP数据"""
        stream_id = event.stream_id
        
        if stream_id not in self._handlers:
            logger.warning(f"Received data for unknown stream {stream_id}")
            return
        
        handler = self._handlers[stream_id]
        handler.add_data(event.data, event.stream_ended)
        
        # 如果流结束，处理请求
        if event.stream_ended:
            asyncio.create_task(self.process_request(stream_id))

    async def process_request(self, stream_id: int):
        """处理完整的HTTP请求"""
        if stream_id not in self._handlers:
            return
        
        handler = self._handlers[stream_id]
        
        try:
            await handler.handle_request(self, stream_id)
        except Exception as e:
            logger.error(f"Error processing request for stream {stream_id}: {e}")
        finally:
            # 清理处理器
            del self._handlers[stream_id]

async def run_server(
    host: str = "0.0.0.0",
    port: int = 6121,
    certificate: str = "certpath/cert.pem",
    private_key: str = "certpath/priv.key",
):
    """运行HTTP/3服务器"""
    
    # 检查证书文件是否存在
    if not os.path.exists(certificate) or not os.path.exists(private_key):
        logger.error(f"Certificate files not found: {certificate}, {private_key}")
        logger.info("Please generate certificates first")
        return

    # 配置QUIC
    configuration = QuicConfiguration(
        alpn_protocols=H3_ALPN,
        is_client=False,
        max_datagram_frame_size=65536,
    )
    
    configuration.load_cert_chain(certificate, private_key)

    # 启动服务器
    server = await serve(
        host=host,
        port=port,
        configuration=configuration,
        create_protocol=Http3ServerProtocol,
    )
    
    logger.info(f"HTTP/3 server listening on https://{host}:{port}")
    
    try:
        await asyncio.Future()  # 永久运行
    except KeyboardInterrupt:
        logger.info("Received interrupt signal, shutting down...")
    finally:
        server.close()

def signal_handler():
    """处理退出信号"""
    logger.info("Shutting down server...")
    # 清理模拟数据存储（可选）
    mock_data_store.clear()
    logger.info("Mock data store cleared")
    sys.exit(0)

if __name__ == "__main__":
    # 设置信号处理
    signal.signal(signal.SIGINT, lambda s, f: signal_handler())
    signal.signal(signal.SIGTERM, lambda s, f: signal_handler())
    
    # 运行服务器
    asyncio.run(run_server())